File: Leetcode/LeetCode_75/Maximum Product Subarray/Maximum Product Subarray.js
function maxProduct(nums) {
  let maxProduct = Math.max(...nums);
  let currentMin = 1, currentMax = 1;

  for (let i = 0; i < nums.length; i++) {
      // debugger;
      let num = nums[i];
      
      if (num === 0) {
          currentMin = 1;
          currentMax = 1;
          continue;
      }
      
      let temp = currentMax * num;
      currentMax = Math.max(temp, num * currentMin, num);
      currentMin = Math.min(temp, num * currentMin, num);

      maxProduct = Math.max(currentMax, maxProduct);
  }

  return maxProduct;
}

// Example usage:
let nums = [2, 3, -2, 4];
console.log(maxProduct(nums)); // Output: 6





function maxProduct2(nums) {
  if (nums.length === 0) return 0;

  let maxProduct = nums[0];
  let minProduct = nums[0];
  let result = nums[0];

  let objTrack = {
    'currentNum': [],
    'maxProductBefore': [],
    'minProductBefore': [],
    'max*curr': [],
    'min*curr': [],
    'maxProductAfter': [],
    'minProductAfter': [],
    'result': []
  }

  for (let i = 1; i < nums.length; i++) {
      let currentNum = nums[i];
      objTrack['currentNum'].push(currentNum);
      objTrack['maxProductBefore'].push(maxProduct);
      objTrack['minProductBefore'].push(minProduct);
      objTrack['max*curr'].push(maxProduct * currentNum);
      objTrack['min*curr'].push(minProduct * currentNum);
      
      // If the current number is negative, swap maxProduct and minProduct
      if (currentNum < 0) {
          [maxProduct, minProduct] = [minProduct, maxProduct];
      }
      // Calculate max and min products for the current number
      maxProduct = Math.max(currentNum, maxProduct * currentNum);
      minProduct = Math.min(currentNum, minProduct * currentNum);

      // Update the result with the maximum product found so far
      result = Math.max(result, maxProduct);
      objTrack['maxProductAfter'].push(maxProduct);
      objTrack['minProductAfter'].push(minProduct);
      objTrack['result'].push(result);
  }
  console.table(objTrack);
  return result;
}

// Example usage:
let nums2 = [2, 3, -2, 4];
console.log(maxProduct(nums2)); // Output: 6

File: Leetcode/LeetCode_75/Maximum Product Subarray/Maximum Product Subarray.rb
# 152. Maximum Product Subarray
# Solved
# Medium
# Topics
# Companies
# Given an integer array nums, find a
# subarray
#  that has the largest product, and return the product.

# The test cases are generated so that the answer will fit in a 32-bit integer.



# Example 1:

# Input: nums = [2,3,-2,4]
# Output: 6
# Explanation: [2,3] has the largest product 6.
# Example 2:

# Input: nums = [-2,0,-1]
# Output: 0
# Explanation: The result cannot be 2, because [-2,-1] is not a subarray.


# Constraints:

# 1 <= nums.length <= 2 * 104
# -10 <= nums[i] <= 10
# The product of any prefix or suffix of nums is guaranteed to fit in a 32-bit integer.

# https://www.youtube.com/watch?v=lXVy6YWFcRM&list=PLot-Xpze53ldVwtstag2TL4HQhAnC8ATf&index=6
# neetcode

# https://www.youtube.com/watch?v=i268ygVolps
# another guy Mayowa Daniel

# @param {Integer[]} nums
# @return {Integer}
def max_product(nums)
  max_product = nums.max
  current_min, current_max = 1, 1

  nums.each_with_index do |num, i|
      if num == 0
          current_min, current_max = 1, 1
          next
      end
      tmp = current_max * num
      current_max = [tmp, num * current_min, num].max
      current_min = [tmp, num * current_min, num].min

      max_product = [current_max, max_product].max
  end
  max_product
end
